Add docs page for metadata (#241)
memgraph/docs: docs/reference-guide/metadata.md

---
id: metadata
title: Metadata
sidebar_label: Metadata
---

Bolt protocol specifies additional data that can be sent along with the
requested results. Such data is called metadata and can be divided into two
groups:
 - Query Statistics,
 - Notifications.

Both of these metadata can be accessed through `summary` map that is being sent
along with the results of the query. Query statistics will be under `stats` key,
and notifications under `notifications` key.

## Query Statistics

Query statistics will be sent whenever a user performs a query that will effect
data in any way. In other words we will track the quantity of these changes
throughout the query execution and report it back to the user.

The structure of statistics is a map of string keys and integer values. Data
that is being tracked:

 - `nodes-created`
 - `nodes-deleted`
 - `relationships-created`
 - `relationships-deleted`
 - `labels-added`
 - `labels-removed`
 - `properties-set`

> 

Reviewer: Please add a sentence or two about the differences between these counters and the triggers.

Reviewer: Ohh, actually I was thinking about a bit different thing. When a user executes a query like `CREATE (n) DELETE n;`, then both `nodes-created` and `nodes-deleted` will be `1`, but no trigger will be triggered, because the triggers accumulates the changes throughout a transaction, while these counters aren't. In my opinion this might be a bit confusing for the users.

But your comment is also true and important, I just didn't think about it before, so I think this is the first time when miscommunication resulted in something better :smile:

Author: Aha ok that makes sense, I will use your example to communicate that better

## Notifications

Notifications will be sent whether we want to confirm the results of query or

> want to want the user of possible wrong usage. To understand what kind of

Reviewer: ```suggestion
want to notify the user about possible wrong usage. To understand what kind of
```

message notification can send let's look at the structure of notification:
```
{
    "severity": "...",
    "code: "...",
    "title": "...",
    "description": "..."
}

> ```

Reviewer: I might use a list/table to name these fields and describe their meaning.

---
id: metadata
title: Metadata
sidebar_label: Metadata
---

Bolt protocol specifies additional data that can be sent along with the
requested results. Such data is called metadata and can be divided into two
groups:
 - Query Statistics,
 - Notifications.

Both of these metadata can be accessed through `summary` map that is being sent
along with the results of the query. Query statistics will be under `stats` key,
and notifications under `notifications` key.

## Query Statistics

Query statistics will be sent whenever a user performs a query that will effect
data in any way. In other words we will track the quantity of these changes
throughout the query execution and report it back to the user.

The structure of statistics is a map of string keys and integer values. Data
that is being tracked:

 - `nodes-created`
 - `nodes-deleted`
 - `relationships-created`
 - `relationships-deleted`
 - `labels-added`
 - `labels-removed`
 - `properties-set`

This data will refer only to the changes done by query, not triggers changes will

> not affect these values.

Reviewer 2: This is wrong actually because `BEFORE COMMIT` will run trigger statements as part of the transaction so it should affect stats.

Author: Hmm it didn't affected the stats

Author: I've tested both BEFORE and AFTER commit neither of these shows up in stats, only the query changes.

> If the query changes and deletes nodes like this `CREATE (n) DELETE n;` and
> leaves memgraph in the same state as before. The results will be 1 for both
> `nodes-created` and `nodes-deleted` but the triggers will not be triggered since
> there were no changes across the transaction.

Reviewer: What about this?

```suggestion
:::caution Differences compared to triggers

It is possible that after executing a query some of these counters are not
zero, however the regarding triggers are not triggered. The reason for that is
triggers are only  triggered when there is a difference between the starting
and ending state, while the counters are also counting the not permanent
changes.

For example if the query changes and deletes nodes like
`CREATE (n) DELETE n;`, then it leaves Memgraph in the same state as before.
The value will be 1 for both `nodes-created` and `nodes-deleted`, but the
triggers will not be triggered since there were no changes across the
transaction.

:::
```

I think it gives a bit more general picture about how the counters and triggers differ while keeping the example for providing something concrete

Edit: I just saw the note from @antonio2368 and it makes this a bit outdated, but as I already wrote this, I think it is better to publish it.

---
id: metadata
title: Metadata
sidebar_label: Metadata
---

Bolt protocol specifies additional data that can be sent along with the
requested results. Such data is called metadata and can be divided into two
groups:

>  - Query Statistics,

Reviewer 3: ```suggestion
 - Query Statistics
```

>  - Notifications.

Reviewer 3: ```suggestion
 - Notifications
```

Both of these metadata can be accessed through `summary` map that is being sent
along with the results of the query. Query statistics will be under `stats` key,
and notifications under `notifications` key.

## Query Statistics

> Query statistics will be sent whenever a user performs a query that will effect

Reviewer 3: ```suggestion
Query statistics will be sent whenever a user executes a query that will affect
```

Reviewer 3: ```suggestion
Query statistics will be sent whenever a user performs a query that will affect
```

Reviewer 3: Either of the two :)

data in any way. In other words we will track the quantity of these changes
throughout the query execution and report it back to the user.

The structure of statistics is a map of string keys and integer values. Data
that is being tracked:

 - `nodes-created`
 - `nodes-deleted`
 - `relationships-created`
 - `relationships-deleted`
 - `labels-added`
 - `labels-removed`
 - `properties-set`

> This data will refer only to the changes done by query, thus changes made in

Reviewer 3: ```suggestion
This data will refer only to the changes done by the query, thus changes made in
```

not affect these values.

:::caution Differences compared to triggers

It is possible that after executing a query some of these counters are not
zero, however the regarding triggers are not triggered. The reason for that is
triggers are only  triggered when there is a difference between the starting
and ending state, while the counters are also counting the not permanent
changes.

> For example if the query changes and deletes nodes like
> `CREATE (n) DELETE n;`, then it leaves Memgraph in the same state as before.
> The value will be 1 for both `nodes-created` and `nodes-deleted`, but the
> triggers will not be triggered since there were no changes across the
> transaction.

Reviewer: ```suggestion
For example if the query creates and deletes nodes like
`CREATE (n) DELETE n;`, then it leaves Memgraph in the same state as before.
The value will be 1 for both `nodes-created` and `nodes-deleted`, but the
triggers will not be triggered since there is no difference between in
states before and after the query is executed.
```

Just to eliminate the "transaction" word from the explanation, because `BEFORE` triggers are run in the transaction.
